src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -1953,16 +2593,152 @@ void dds_stream_extract_keyhash (dds_istream_t * __restrict is, dds_keyhash_t *
+  }
+  else
+  {
+    if (desc->flagset & DDS_TOPIC_CONTAINS_UNION)
+    {
+      /* Switching union cases causes big trouble if some cases have sequences or strings,
+         and other cases have other things mapped to those addresses.  So, pretend to be
+         nice by freeing whatever was allocated, then clearing all memory.  This will
+         make any preallocated buffers go to waste, but it does allow reusing the message
+         from read-to-read, at the somewhat reasonable price of a slower deserialization
+         and not being able to use preallocated sequences in topics containing unions. */
+      dds_stream_free_sample (data, desc->ops.ops);
+      memset (data, 0, desc->size);
+    }
+    (void) dds_stream_read (is, data, desc->ops.ops);
+  }
+}
+
+static void dds_stream_read_key_impl (dds_istream_t * __restrict is, char * __restrict sample, const uint32_t *insnp, uint16_t key_offset_count, const uint32_t * key_offset_insn)
+{
+  char *dst = sample + insnp[1];
+  assert (insn_key_ok_p (*insnp));
+  switch (DDS_OP_TYPE (*insnp))
+  {
+    case DDS_OP_VAL_1BY: *((uint8_t *) dst) = dds_is_get1 (is); break;
+    case DDS_OP_VAL_2BY: *((uint16_t *) dst) = dds_is_get2 (is); break;
+    case DDS_OP_VAL_4BY: case DDS_OP_VAL_ENU: *((uint32_t *) dst) = dds_is_get4 (is); break;
+    case DDS_OP_VAL_8BY: *((uint64_t *) dst) = dds_is_get8 (is); break;
+    case DDS_OP_VAL_STR: *((char **) dst) = dds_stream_reuse_string (is, *((char **) dst)); break;
+    case DDS_OP_VAL_BST: (void) dds_stream_reuse_string_bound (is, dst, insnp[2], false); break;
+    case DDS_OP_VAL_BSP: *((char **) dst) = dds_stream_reuse_string_bound (is, dst, insnp[2], true); break;
+    case DDS_OP_VAL_ARR: dds_is_get_bytes (is, dst, insnp[2], get_type_size (DDS_OP_SUBTYPE (*insnp))); break;
+    case DDS_OP_VAL_SEQ: case DDS_OP_VAL_UNI: case DDS_OP_VAL_STU: abort (); break;
+    case DDS_OP_VAL_EXT:
+    {
+      assert (key_offset_count > 0);
+      const uint32_t *jsr_ops = insnp + DDS_OP_ADR_JSR (insnp[2]) + *key_offset_insn;
+      dds_stream_read_key_impl (is, dst, jsr_ops, --key_offset_count, ++key_offset_insn);
+      break;
+    }
+  }
+}
+
+void dds_stream_read_key (dds_istream_t * __restrict is, char * __restrict sample, const struct ddsi_sertype_default * __restrict type)
+{
+  const struct ddsi_sertype_default_desc *desc = &type->type;
+  for (uint32_t i = 0; i < desc->keys.nkeys; i++)
+  {
+    const uint32_t *op = desc->ops.ops + desc->keys.keys[i];
+    switch (DDS_OP (*op))
+    {
+      case DDS_OP_KOF: {
+        uint16_t n_offs = DDS_OP_LENGTH (*op);
+        dds_stream_read_key_impl (is, sample, desc->ops.ops + op[1], --n_offs, op + 2);
+        break;
+      }
+      case DDS_OP_ADR: {
+        dds_stream_read_key_impl (is, sample, op, 0, NULL);
+        break;
+      }
+      default:
+        abort ();
+        break;
+    }
+  }
+}
+
+static inline void dds_stream_swap_insitu (void * __restrict vbuf, uint32_t size, uint32_t num)
```

Comment:
This needs to be renamed if it must exist as a no-op function on big-endian machines. Perhaps `dds_stream_to_bigendian_insitu`?

Reply:
This function is used when writing keys in native endianness (and therefore is a no-op), but I can imagine that the name is somewhat confusing in that case. So I'll change its name to `dds_stream_swap_if_needed_insitu` and add a comment to explain its use.